google/cloud/sql/connector/rate_limiter.py

```diff
@@ -0,0 +1,126 @@
+"""
+Copyright 2021 Google LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+  https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+import asyncio
+import math
+import time
+from typing import Optional, Type
+from types import TracebackType
+
+
+class AsyncRateLimiter(object):
+    """
+    An asyncio-compatible rate limiter which uses the Leaky Bucket algorithm
+    (https://en.wikipedia.org/wiki/Leaky_bucket) to limit the number of function
+    calls over a time interval using an event queue.
+
+    :type burst_size: int
+    :param: burst_size:
+        The max number of function calls that can occur at the same time
```

Comment:
This is the bucket capacity right? So not necessarily the number of calls that can occur at the same time, but more so the max the bucket can fill up to when not being used?

Reply:
Ok, so this is my mistake - I actually gave you the wrong name of the algorithm. I said "Leaky Bucket", but I really meant "Token Bucket": https://en.wikipedia.org/wiki/Token_bucket

I think the token bucket is preferred because it'll let us "save up tokens" when we aren't using, and let us burst slightly if we need too. 

WDYT about using a token bucket instead?

Reply:
I actually had this using the token bucket algorithm originally (which is why those comments about bursts were left in, I forgot to remove them). I switched to leaky bucket because I thought it's better for our use case, since I figured we didn't want a bunch of refresh calls occurring at around the same time (i.e. a refresh starts before another one can complete). I'm not sure if bursting is useful for our use case, or if it might actually end up causing issues.

Reply:
I updated it to use a token bucket again, but I'm still not convinced it's the better option.

Reply:
I think if we can figure out the other thread we have going on (how to cancel a not-yet-started task), that we won't need to worry about more than one task going at the same time. 

The reason I think token is preferred over leaky is because leaky limits outgoing traffic, but has a queue. We don't really ever need to queue attempts, we can stop once a perform refresh is successful. Token on the other hand, makes it OK to perform 2-3 refreshes in a row that helps weed out transient errors, but ensures that amortized over a couple of minutes, our consumption will work as expected.

Reply:
Could we switch to using one of [these methods](https://docs.python.org/3/library/asyncio-eventloop.html#scheduling-delayed-callbacks) instead of the `schedule_refresh` coroutine to schedule tasks? The `asyncio.TimerHandle` lets us schedule a callback at a specific time and then cancel it. So when scheduling the next refresh, we could do something like:
```
self._next = loop.call_later(self.seconds_until_refresh(), lambda: loop.create_task(self.perform_refresh()))
```

Reply:
That seems reasonable to me, as long as we can tell if `cancel` was successful or not.

Reply:
I ended up going with the event approach that you suggested in the other thread. I also added a test to confirm that force_refresh works.